Code cleanup, direct opaque data buffer getter on databuffer, debugging for device data

agibsonccc opened the pull request:
## What changes were proposed in this pull request?
Adds debugging utilities for printing device and host buffers
Add direct opaque buffer getter to base databuffer allowing for direct usage of the interop buffer
for manipulation in c++
Code clean up (removes commented code)

(Please fill in changes proposed in this fix)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

## Quick checklist

The following checklist helps ensure your PR is complete:

- [ X] Eclipse Contributor Agreement signed, and signed commits - see [IP Requirements](https://deeplearning4j.org/eclipse-contributors) page for details
- [X] Reviewed the [Contributing Guidelines](https://github.com/eclipse/deeplearning4j/blob/master/CONTRIBUTING.md) and followed the steps within.
- [X] Created tests for any significant new code additions.
- [X ] Relevant tests for your changes are passing.

agibsonccc committed:
Remove old deeplearning4j.org links
Remove old readmes

agibsonccc committed:
Remove old deeplearning4j.org links
Remove old readmes
integrate performance tracker with unified profiler
Update old profiler configuration checks to use new profiler configuration
Fix synchronization issue with host pointers when overrwriting a primary buffer on cuda backend

agibsonccc committed:
Remove old deeplearning4j.org links
Remove old readmes
integrate performance tracker with unified profiler
Update old profiler configuration checks to use new profiler configuration
Fix synchronization issue with host pointers when overrwriting a primary buffer on cuda backend
Remove logging

agibsonccc committed:
Remove old deeplearning4j.org links
Remove old readmes
integrate performance tracker with unified profiler
Update old profiler configuration checks to use new profiler configuration
Fix synchronization issue with host pointers when overrwriting a primary buffer on cuda backend
Remove logging
Fix up batch gemmm parameters, add extra validation for batch gemm before op executes

agibsonccc committed:
Add  new getNumElements for data buffer
Add printing of pointers for nativeops
Fix up confusion matrix

agibsonccc committed:
Fix save npy signatures
Add data type validation for invalid data types
Remove extra prints

treo commented on libnd4j/include/ops/declarable/helpers/cpu/confusion.cpp:
```diff
@@ -38,6 +38,8 @@ static void _confusionFunctor(NDArray* labels, NDArray* predictions, NDArray* we
       T curr = arrs.at(label)->e<T>(pred);
       arrs.at(label)->p<T>(pred, curr + value);
+      auto idx = shape::getIndexOffset(pred, arrs[0]->shapeInfo());
+      sd_printf("Label %d pred %d idx %d weight val %f\n",label,pred,idx,value);
```
I guess that is a left over from debugging?

treo commented on libnd4j/include/ops/declarable/helpers/cuda/confusion.cu:
```diff
@@ -31,100 +31,128 @@ namespace helpers {
+ const T * sourceCast = reinterpret_cast<T const*>(source);
+ for (int t = tid; t < bufferLength; t += step) {
+   destination[t] = static_cast<sd::LongType>(sourceCast[t]);
+   printf("New value at value %d is %d source was %d\n",t,destination[t],static_cast<sd::LongType>(reinterpret_cast<T const*>(source)[t]));
```
left over from debugging?

treo commented on libnd4j/include/ops/declarable/helpers/cuda/confusion.cu:
```diff
@@ -31,100 +31,128 @@ namespace helpers {
+ for (int t = tid; t < bufferLength; t += step) {
+   auto label = labelsBuffer[t];     //->e<sd::LongType>(j);
+   auto pred = predictionBuffer[t];  //->e<sd::LongType>(j);
+   printf("Label is %d pred is %d\n",label,pred);
```
left over from debugging?

treo commented on libnd4j/include/ops/declarable/helpers/cuda/confusion.cu:
```diff
@@ -31,100 +31,128 @@ namespace helpers {
+
+ auto pack = sd::ConstantTadHelper::getInstance().tadForDimensions(output->shapeInfo(), 1);
+ labels->printBuffer("Labels host buffer");
+ predictions->printBuffer("Predictions host buffer");
```
left over from debugging?

treo commented on libnd4j/include/ops/declarable/helpers/cuda/confusion.cu:
```diff
@@ -31,100 +31,128 @@ namespace helpers {
+ printBuffers<<<256, 512, 1024, *stream>>>(labelsLongBuffer, labels->lengthOf());
+ manager.synchronize();
+ sd_printf("Printing predictions device buffer\n",0);
+ printBuffers<<<256, 512, 1024, *stream>>>(predictionLongBuffer,  labels->lengthOf());
```
left over from debugging?

treo commented on nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/nativeblas/NativeOps.java:
```diff
@@ -26,7 +26,20 @@
+     * @param buffer
+     */
+    void printDeviceBuffer(org.nd4j.nativeblas.OpaqueDataBuffer buffer);
+
```
Indent looks wonky.

agibsonccc committed:
Fix save npy signatures
Add data type validation for invalid data types
Remove extra prints
Fix formatting nits

agibsonccc committed:
Fix save npy signatures
Add data type validation for invalid data types
Remove extra prints
Fix formatting nits

agibsonccc committed:
Merge branch 'master' into ag_more_test_fixes

agibsonccc committed:
Fix save npy signatures
Add data type validation for invalid data types
Remove extra prints
Fix formatting nits
fix up merge conflict issues

agibsonccc committed:
Fix save npy signatures
Add data type validation for invalid data types
Remove extra prints
Fix formatting nits
fix up merge conflict issues
Fix numpy cnpy conflict

agibsonccc committed:
Fix save npy signatures
Add data type validation for invalid data types
Remove extra prints
Fix formatting nits
fix up merge conflict issues
Fix numpy cnpy conflict

agibsonccc committed:
Fix save npy signatures
Add data type validation for invalid data types
Remove extra prints
Fix formatting nits
fix up merge conflict issues
Fix numpy cnpy conflict
Remove extra print statements
